README.md
# searchConsoleR

[![CRAN](http://www.r-pkg.org/badges/version/searchConsoleR)](https://CRAN.R-project.org/package=searchConsoleR)
[![Travis-CI Build Status](https://travis-ci.org/MarkEdmondson1234/searchConsoleR.svg?branch=master)](https://travis-ci.org/MarkEdmondson1234/searchConsoleR)

R interface with Google Search Console (formally Google Webmaster Tools) API v3.

## Setup Guide

Install dependency `googleAuthR` from CRAN:
```r
install.packages("googleAuthR")
library(googleAuthR)
```

Install `searchConsoleR` 0.3.0 from CRAN:
```r
install.packages("searchConsoleR")
library(searchConsoleR)
```

If you want the development version of `searchConsoleR` on Github:

```r
devtools::install_github("MarkEdmondson1234/searchConsoleR")
library(searchConsoleR)
```

## Shiny Compatible
Authentication can be done locally or within a Shiny app. See a very bare bones example here: https://mark.shinyapps.io/searchConsoleRDemo/

## Info Links

[Google Search Console](http://www.google.com/webmasters/tools/)

[Search Console v3 API docs](https://developers.google.com/webmaster-tools/)

## Function Quick Guide

### Search analytics
* `search_analytics()` - download Google SEO data into an R dataframe.

### Website admin
* `list_websites()` - list websites in your Google Search Console.
* `add_website()` - add a website to your Google Search Console.
* `delete_website()` - delete a website from your Google Search Console.

### Sitemaps
* `list_sitemaps()` - list sitemaps recognised in Google Search Console.
* `add_sitemap()` - add sitemap URL location to Google Search Console.
* `delete_sitemap()` - remove sitemap URL location in Google Search Console.

### Error listings
* `crawl_errors()` - list various types of crawl errors googlebot has found.
* `list_crawl_error_samples()` - get a list of example URLs with errors.
* `error_sample_url()` - show details about an example URL error (for example, links to a 404 URL)
* `fix_sample_url()` - mark a URL as fixed.

### Authentication functions from googleAuthR

* `scr_auth()` - main authentication function. Works locally and within a Shiny environment.

## Work flow

Work flow always starts with authenticating with Google:
```r
library(searchConsoleR)
scr_auth()
```

Your browser window should open up and go through the Google sign in OAuth2 flow. Verify with a user that has Search Console access to the websites you want to work with.

Check out the documentation of any function for a guide on what else can be done.
```r
?searchConsoleR
```

If you authenticate ok, you should be able to see a list of your websites in the Search Console via:

```r
sc_websites <- list_websites()
sc_websites
```

We'll need one unique ```sc_websites$siteUrl``` for the majority of the other functions.

Most people will find the Search Analytics most useful.  All methods from the web interface are available.  

Here is an example query, which downloads the top 100 rows of queries per page for the month of July 2015, for United Kingdom desktop web searches:

```r
gbr_desktop_queries <- 
    search_analytics("http://example.com", 
                     "2015-07-01", "2015-07-31", 
                     c("query", "page"), 
                     dimensionFilterExp = c("device==DESKTOP","country==GBR"), 
                     searchType="web", rowLimit = 100)
```

For a lot more details see: 
```r
?search_analytics
```

### Batching

You can get more than the standard 5000 rows via batching.  There are two methods available, one via a API call per date, the other using the APIs `startRow` parameter.

The date method gets more impressions for 0 click rows, the batch method is quicker but gets just rows with clicks. 

Specify a `rowLimit` when batching - if using method `byDate` this will be the limit it fetches per day, and currently needs to be over 5000 to work. [(Issue #17 will fix this)](https://github.com/MarkEdmondson1234/searchConsoleR/issues/17).

```r
test0 <- search_analytics("http://www.example.co.uk", 
                          dimensions = c("date","query","page","country"), 
                          rowLimit = 200000, 
                          walk_data = "byBatch")
Batching data via method: byBatch

### test0 has 13063 rows

test <- search_analytics("http://www.example.co.uk", 
                         dimensions = c("date","query","page","country"), 
                         walk_data = "byDate")
Batching data via method: byDate

### test has 419957 rows

> sum(test0$clicks)
[1] 12866
> sum(test$clicks)
[1] 12826
> sum(test$impressions)
[1] 1420217
> sum(test0$impressions)
[1] 441029
> 

```

## Demo script

Here is an example for downloading daily data and exporting to .csv

```r
## A script to download and archive Google search analytics
##
## Demo of searchConsoleR R package.
##
## Version 1 - 10th August 2015
##
## Mark Edmondson (http://markedmondson.me)

library(searchConsoleR)

## change this to the website you want to download data for. Include http
website <- "http://copenhagenish.me"

## data is in search console reliably 3 days ago, so we donwnload from then
## today - 3 days
start <- Sys.Date() - 3
## one days data, but change it as needed
end <- Sys.Date() - 3 

## what to download, choose between date, query, page, device, country
download_dimensions <- c('date','query')

## what type of Google search, choose between 'web', 'video' or 'image'
type <- c('web')

## other options available, check out ?search_analytics in the R console

## Authorize script with Search Console.  
## First time you will need to login to Google,
## but should auto-refresh after that so can be put in 
## Authorize script with an account that has access to website.
scr_auth()

## first time stop here and wait for authorisation

## get the search analytics data
data <- search_analytics(siteURL = website, 
                         startDate = start, 
                         endDate = end, 
                         dimensions = download_dimensions, 
                         searchType = type)

## do stuff to the data
## combine with Google Analytics, filter, apply other stats etc.

## write a csv to a nice filename
filename <- paste("search_analytics",
                  Sys.Date(),
                  paste(download_dimensions, collapse = "",sep=""),
                  type,".csv",sep="-")

write.csv(data, filename)
```

## The dimensionFilterExp parameter

This parameter is used in search_analytics to filter the result.

Filter using this format: ```filter operator expression```

Filter can be one of:

* `country`,
* `device`
* `page`
* `query`

Operator can be one of ```~~, ==, !~, !=``` where the symbols mean:

* `~~` : 'contains',
* `==` : 'equals',
* `!~` : 'notContains',
* `!=` : 'notEquals'

Expression formatting:

* for ```page``` or ```query``` is free text.
* for ```country``` must be the three letter country code as per the [the ISO 3166-1 alpha-3](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3) standard. e.g. USA, GBR = United Kingdom, DNK = Denmark
* for ```device``` must be one of:  'MOBILE', 'DESKTOP' or 'TABLET'

You can have multiple ```AND``` filters by putting them in a character vector.  The below looks for desktop searches in the United Kingdom, not showing the homepage and not including queries containing 'brandterm'.

```
c("device==DESKTOP","country==GBR", "page!=/home", "query!~brandterm")
```

```OR``` filters aren't yet supported in the API.

## Using your own Google API project 

As default `searchConsoleR` uses its own Google API project to grant requests, but if you want to use your own keys:

1. Set up your project in the [Google API Console](https://developers.google.com/identity/sign-in/web/devconsole-project) to use the search console v3 API.

### For local use
2. Click 'Create a new Client ID', and choose "Installed Application".
3. Note your Client ID and secret.
4. Modify these options after `searchConsoleR` has been loaded:
  + `options("searchConsoleR.client_id" = "YOUR_CLIENT_ID")`
  + `options("searchConsoleR.client_secret" = "YOUR_CLIENT_SECRET")`

### For Shiny use
2. Click 'Create a new Client ID', and choose "Web Application".
3. Note your Client ID and secret.
4. Add the URL of where your Shiny app will run, as well as your local host for testing including a port number.  e.g. https://mark.shinyapps.io/searchConsoleRDemo/ and http://127.0.0.1:4624
5. In your Shiny script modify these options:
  + `options("searchConsoleR.webapp.client_id" = "YOUR_CLIENT_ID")`
  + `options("searchConsoleR.webapp.client_secret" = "YOUR_CLIENT_SECRET")`
6. Run the app locally specifying the port number you used e.g. `shiny::runApp(port=4624)`
7. Or deploy to your Shiny Server that deploys to web port (80 or 443).


news.md
# 0.4.0.9000

* Return an empty data.frame() if no websites available for the authenticated user instead of an error
* Allow a JSON service key authentication with `scr_auth(json="your-key.json")`
* Make sure ctr and position columns are always float numerics class.
* Add inspection API (#73) via `inspection()`)
* Add regex support for `search_analytics()` via `**` and `!*`

# 0.4.0

* Support up to 25,000 rows per API call (#44)
* Change paging for `byBatch` to early exit if there is no data
* Support `sc-domain` for URLs (#51)
* Remove `crawl_errors()` call as no longer in API (#52)
* Rely on new `googleAuthR>1.0.0`

# 0.3.0

* Update authentication options for `scr_auth()` to include auto-authentication etc. from googleAuthR `0.6.2`
* Added `searchAppearance` as a dimension option in `search_analytics()`
* Remove warning if data is more than 90 days old as it will soon allow 12 months (woop)

# 0.2.1

* Add batching per the new API feature to go over 5000 rows. (#12)

# 0.2.0

* Return an empty dataframe of NAs if no results in fetch instead of NULL
* Include android-app check (#7)
* Add `walk_data` parameter to `search_analytics` to get more data
* Set default start and end dates in `search_analytics` to 93 days ago and 3 days ago respectively.
* Correct bug for error in country code.  Will now return the 'Unknown Region' if not recognised (e.g. `CXX`)
* Add `scr_auth` function that wraps `googleAuthR::gar_auth` so you don't need to load googleAuthR explicitly.

# 0.1.2

* Move to using googleAuthR for authentication backend.

### 0.1.1
* Change `search_analytics()` so if no dimensions will still return data, instead of NULL


R/getData.R
#' Query search traffic keyword data
#'
#' @description Download your Google SEO data.
#'
#' @param siteURL The URL of the website you have auth access to.
#' @param startDate Start date of requested range, in YYYY-MM-DD.
#' @param endDate End date of the requested date range, in YYYY-MM-DD.
#' @param dimensions Zero or more dimensions to group results by:
#'      \code{"date", "country", "device", "page" , "query" or "searchAppearance"}
#' @param searchType Search type filter, default 'web'.
#' @param dimensionFilterExp A character vector of expressions to filter.
#'      e.g. \code{("device==TABLET", "country~~GBR", "query**^a")}
#' @param aggregationType How data is aggregated.
#' @param rowLimit How many rows to fetch.  Ignored if \code{walk_data} is "byDate"
#' @param prettyNames If TRUE, converts SO 3166-1 alpha-3 country code to full name and
#'   creates new column called countryName.
#' @param walk_data Make multiple API calls. One of \code{("byBatch","byDate","none")}
#'
#' @return A dataframe with columns in order of dimensions plus metrics, with attribute "aggregationType"
#'
#' @seealso Guide to Search Analytics: \url{https://support.google.com/webmasters/answer/6155685}
#'   API docs: \url{https://developers.google.com/webmaster-tools/v3/searchanalytics/query}
#' @export
#'
#' @details
#'  \strong{startDate}: Start date of the requested date range, in YYYY-MM-DD format,
#'    in PST time (UTC - 8:00). Must be less than or equal to the end date.
#'    This value is included in the range.
#'
#'  \strong{endDate}: End date of the requested date range, in YYYY-MM-DD format,
#'    in PST time (UTC - 8:00). Must be greater than or equal to the start date.
#'    This value is included in the range.
#'
#'  \strong{dimensions}: [Optional] Zero or more dimensions to group results by.
#'       \itemize{
#'         \item 'date'
#'         \item 'country'
#'         \item 'device'
#'         \item 'page'
#'         \item 'query'
#'         \item 'searchAppearance' (can only appear on its own)
#'       }
#'  The grouping dimension values are combined to create a unique key
#'    for each result row. If no dimensions are specified,
#'    all values will be combined into a single row.
#'    There is no limit to the number of dimensions that you can group by apart from \code{searchAppearance} can only be grouped alone.
#'    You cannot group by the same dimension twice. 
#'    
#'    Example: \code{c('country', 'device')}
#'    
#'
#'  \strong{dimensionFilterExp}:
#'  Results are grouped in the order that you supply these dimensions.
#'  dimensionFilterExp expects a character vector of expressions in the form:
#'   ("device==TABLET", "country~~GBR", "dimension operator expression")
#'   \itemize{
#'     \item dimension
#'       \itemize{
#'         \item 'country'
#'         \item 'device'
#'         \item 'page'
#'         \item 'query'
#'         \item 'searchAppearance'
#'       }
#'     \item operator
#'       \itemize{
#'         \item '~~' meaning 'contains'
#'         \item '==' meaning 'equals'
#'         \item '!~' meaning 'notContains'
#'         \item '!=' meaning 'notEquals'
#'         \item '**' meaning 'includingRegex'
#'         \item '!*' meaning 'excludingRegex'
#'       }
#'
#'     \item expression
#'        \itemize{
#'          \item country: an ISO 3166-1 alpha-3 country code.
#'          \item device: 'DESKTOP','MOBILE','TABLET'.
#'          \item page: not checked, a string in page URLs without hostname.
#'          \item query: not checked, a string in keywords.
#'          \item searchAppearance: 'AMP_BLUE_LINK', 'RICHCARD'
#'
#'        }
#'   }
#'
#'
#'  \strong{searchType}: [Optional] The search type to filter for. Acceptable values are:
#'  \itemize{
#'    \item "web": [Default] Web search results
#'    \item "image": Image search results
#'    \item "video": Video search results
#'    \item "news": "News" tab in Google Search
#'    \item "googleNews": Results from news.google.com. Doesn't include results from the "news" tab in Google Search
#'    \item "discover": Discover results
#'  }
#'
#'  \strong{aggregationType}: [Optional] How data is aggregated.
#'  \itemize{
#'    \item If aggregated by property, all data for the same property is aggregated;
#'    \item If aggregated by page, all data is aggregated by canonical URI.
#'    \item If you filter or group by page, choose auto; otherwise you can aggregate either by property or by page, depending on how you want your data calculated;
#'  }
#'    See the API documentation to learn how data is calculated differently by site versus by page.
#'    Note: If you group or filter by page, you cannot aggregate by property.
#'    If you specify any value other than auto, the aggregation type in the result will match the requested type, or if you request an invalid type, you will get an error.
#'    The API will never change your aggregation type if the requested type is invalid.
#'    Acceptable values are:
#'  \itemize{
#'    \item "auto": [Default] Let the service decide the appropriate aggregation type.
#'    \item "byPage": Aggregate values by URI.
#'    \item "byProperty": Aggregate values by property.
#'  }
#'
#'  \strong{batchType}: [Optional] Batching data into multiple API calls
#' \itemize{
#'   \item byBatch Use the API call to batch
#'   \item byData Runs a call over each day in the date range.
#'   \item none No batching
#'  }
#'
#' @examples
#'
#' \dontrun{
#'
#'    library(searchConsoleR)
#'    scr_auth()
#'    sc_websites <- list_websites()
#'    
#'    default_fetch <- search_analytics("http://www.example.com")
#'
#'    gbr_desktop_queries <-
#'        search_analytics("http://www.example.com",
#'                          start = "2016-01-01", end = "2016-03-01",
#'                          dimensions = c("query", "page"),
#'                          dimensionFilterExp = c("device==DESKTOP", "country==GBR"),
#'                          searchType = "web", rowLimit = 100)
#'
#'    batching <-
#'         search_analytics("http://www.example.com",
#'                          start = "2016-01-01", end = "2016-03-01",
#'                          dimensions = c("query", "page", "date"),
#'                          searchType = "web", rowLimit = 100000,
#'                          walk_data = "byBatch")
#'
#'   }
#' @importFrom googleAuthR gar_api_generator gar_batch_walk gar_api_page
search_analytics <- function(siteURL,
                             startDate = Sys.Date() - 93,
                             endDate = Sys.Date() - 3,
                             dimensions = NULL,
                             searchType = c("web","video","image","news","discover","googleNews"),
                             dimensionFilterExp = NULL,
                             aggregationType = c("auto","byPage","byProperty"),
                             rowLimit = 1000,
                             prettyNames = TRUE,
                             walk_data = c("byBatch","byDate","none")){

  if(!googleAuthR::gar_has_token()){
    stop("Not authenticated. Run scr_auth()", call. = FALSE)
  }

  searchType      <- match.arg(searchType)
  aggregationType <- match.arg(aggregationType)
  walk_data       <- match.arg(walk_data)

  startDate <- as.character(startDate)
  endDate   <- as.character(endDate)

  message("Fetching search analytics for ",
          paste("url:", siteURL,
                "dates:", startDate, endDate,
                "dimensions:", paste(dimensions, collapse = " ", sep=";"),
                "dimensionFilterExp:", paste(dimensionFilterExp, collapse = " ", sep=";"),
                "searchType:", searchType,
                "aggregationType:", aggregationType))

  siteURL <- check.Url(siteURL, reserved=T)

  if(any(is.na(as.Date(startDate, "%Y-%m-%d")), is.na(as.Date(endDate, "%Y-%m-%d")))){
    stop("dates not in correct %Y-%m-%d format. Got these:", startDate, " - ", endDate)
  }

  if(any(as.Date(startDate, "%Y-%m-%d") > Sys.Date()-3, as.Date(endDate, "%Y-%m-%d") > Sys.Date()-3)){
    warning("Search Analytics usually not available within 3 days (96 hrs) of today(",Sys.Date(),"). Got:", startDate, " - ", endDate)
  }

  if(!is.null(dimensions) && !dimensions %in% c('date','country', 'device', 'page', 'query','searchAppearance')){
    stop("dimension must be NULL or one or more of 'date','country', 'device', 'page', 'query', 'searchAppearance'.
         Got this: ", paste(dimensions, sep=", "))
  }

  if(aggregationType %in% c("byProperty") && 'page' %in% dimensions ){
    stop("Can't aggregate byProperty and include page in dimensions.")
  }
  
  # if batching by day, row limits make no sense so we get 5000 per day.
  if(walk_data == "byDate"){
    message("Batching data via method: ", walk_data)
    message("Will fetch up to 25000 rows per day")
    rowLimit <- 25000
  } else if(walk_data == "byBatch"){
    # if batching byBatch, we set to 25000 per API call, repeating API calls
    #   up to the limit you have set
    if(rowLimit > 25000){
      message("Batching data via method: ", walk_data)
      message("With rowLimit set to ", rowLimit ," will need up to [", (rowLimit %/% 25000) + 1, "] API calls")
      rowLimit0 <- rowLimit
      rowLimit <- 25000
    } else {
      # its batched, but we can get all rows in one API call
      walk_data <- "none"
    }
  }

  ## a list of filter expressions
  ## expects dimensionFilterExp like c("device==TABLET", "country~~GBR")
  parsedDimFilterGroup <- lapply(dimensionFilterExp, parseDimFilterGroup)

  body <- list(
    startDate = startDate,
    endDate = endDate,
    dimensions = as.list(dimensions),
    type = searchType,
    dimensionFilterGroups = list(
      list( ## you don't want more than one of these until different groupType available
        groupType = "and", ##only one available for now
        filters = parsedDimFilterGroup
      )
    ),
    aggregationType = aggregationType,
    rowLimit = rowLimit
  )

  search_analytics_g <- gar_api_generator(
    "https://www.googleapis.com/webmasters/v3/",
    "POST",
    path_args = list(sites = "siteURL",
                     searchAnalytics = "query"),
    data_parse_function = parse_search_analytics)
  
  
  # set this here as it may get reset if other googleAuthR packages there
  options(googleAuthR.batch_endpoint = 'https://www.googleapis.com/batch/webmasters/v3')
  
  if(walk_data == "byDate"){

    if(!'date' %in% dimensions){
      warning("To walk data per date requires 'date' to be one of the dimensions. Adding it")
      dimensions <- c("date", dimensions)
    }

    walk_vector <- seq(as.Date(startDate), as.Date(endDate), 1)

    out <- gar_batch_walk(search_analytics_g,
                          walk_vector = walk_vector,
                          gar_paths = list(sites = siteURL),
                          body_walk = c("startDate", "endDate"),
                          the_body = body,
                          batch_size = 1,
                          dim = dimensions)

  } else if(walk_data == "byBatch") {

    ## byBatch uses API batching, but this pulls out less data
    ## 0 impression keywords not included.
    walk_vector <- seq(0, rowLimit0, 25000)
    
    do_it <- TRUE
    i <- 1
    pages <- list()
    while(do_it){
      message("Page [",i,"] of max [", length(walk_vector),"] API calls")
      this_body <- utils::modifyList(body, list(startRow = walk_vector[i]))
      this_page <- search_analytics_g(the_body = this_body, 
                                     list(sites = siteURL),
                                     dim = dimensions)

      if(all(is.na(this_page[[1]]))){
        do_it <- FALSE
      } else {
        message("Downloaded ", nrow(this_page), " rows")
        pages <- rbind(pages, this_page)
      }
     
      i <- i + 1
      if(i>length(walk_vector)){
        do_it <- FALSE
      }

    }
    
    out <- pages

  } else {

    out <-   search_analytics_g(the_body=body,
                                path_arguments=list(sites = siteURL),
                                dim = dimensions)

  }

  out

}


#' Retrieves dataframe of websites user has in Search Console
#'
#' @return a dataframe of siteUrl and permissionLevel
#'
#' @export
#' @family search console website functions
list_websites <- function() {
  
  f <- function(x){
    if(is.null(x$siteEntry)){
      message("No websites found for this authentication")
      return(data.frame())
    }
    
    x$siteEntry
  }

  l <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/sites",
                                      "GET",
                                      data_parse_function = f)
  l()
}

#' Adds website to Search Console
#'
#' @param siteURL The URL of the website to add.
#'
#' @return TRUE if successful, raises an error if not.
#' @family search console website functions
#'
#' @export
add_website <- function(siteURL) {

  siteURL <- check.Url(siteURL, reserved = TRUE)

  aw <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                      "PUT",
                                      path_args = list(sites = "siteURL"))

  aw(path_arguments = list(sites = siteURL))
  TRUE

}

#' Deletes website in Search Console
#'
#' @param siteURL The URL of the website to delete.
#'
#' @return TRUE if successful, raises an error if not.
#' @family data fetching functions
#'
#' @export
#' @family search console website functions
delete_website <- function(siteURL) {

  siteURL <- check.Url(siteURL, reserved = TRUE)


  dw <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                      "DELETE",
                                      path_args = list(sites = "siteURL"))

  dw(path_arguments = list(sites = siteURL))
  TRUE

}

#' Gets sitemap information for the URL supplied.
#'
#' See here for details: https://developers.google.com/webmaster-tools/v3/sitemaps
#'
#' @param siteURL The URL of the website to get sitemap information from. Must include protocol (http://).
#'
#' @return A list of two dataframes: $sitemap with general info and $contents with sitemap info.
#' @family data fetching functions
#'
#' @export
#' @family sitemap admin functions
list_sitemaps <- function(siteURL) {

  siteURL <- check.Url(siteURL, reserved = TRUE)

  ls <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                      "GET",
                                      path_args = list(sites = "siteURL",
                                                       sitemaps = ""),
                                      data_parse_function = parse_sitemaps)

  ls(path_arguments = list(sites = siteURL))

}

#' Submit a sitemap.
#'
#' See here for details: https://developers.google.com/webmaster-tools/v3/sitemaps/submit
#'
#' @param siteURL The URL of the website to delete. Must include protocol (http://).
#' @param feedpath The URL of the sitemap to submit. Must include protocol (http://).
#'
#' @return TRUE if successful, raises an error if not.
#'
#' @export
#' @family sitemap admin functions
add_sitemap <- function(siteURL, feedpath) {

  siteURL <- check.Url(siteURL, reserved = TRUE)
  feedpath <- check.Url(feedpath, reserved = TRUE)

  as <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                      "PUT",
                                      path_args = list(sites = "siteURL",
                                                       sitemaps = "feedpath"))

  as(path_arguments = list(sites = siteURL,
                          sitemaps = feedpath))
  TRUE

}

#' Delete a sitemap.
#'
#' See here for details: https://developers.google.com/webmaster-tools/v3/sitemaps/delete
#'
#' @param siteURL The URL of the website you are deleting the sitemap from. Must include protocol (http://).
#' @param feedpath The URL of the sitemap to delete. Must include protocol (http://).
#'
#' @return TRUE if successful, raises an error if not.
#'
#' @export
#' @family sitemap admin functions
delete_sitemap <- function(siteURL, feedpath) {

  siteURL <- check.Url(siteURL, reserved = TRUE)
  feedpath <- check.Url(feedpath, reserved = TRUE)

  ds <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                      "DELETE",
                                      path_args = list(sites = "siteURL",
                                                       sitemaps = "feedpath"))

  ds(path_arguments = list(sites = siteURL,
                          sitemaps = feedpath))

  TRUE

}

#' Fetch a time-series of Googlebot crawl errors.
#'
#' @description
#' Get a list of errors detected by Googlebot over time.
#' See here for details: https://developers.google.com/webmaster-tools/v3/urlcrawlerrorscounts/query
#'
#' @param siteURL The URL of the website to delete. Must include protocol (http://).
#' @param category Crawl error category. Defaults to 'all'
#' @param platform The user agent type. 'all', 'mobile', 'smartphoneOnly' or 'web'.
#' @param latestCountsOnly Default FALSE. Only the latest crawl error counts returned if TRUE.
#'
#' @return dataframe of errors with $platform $category $count and $timecount.
#'
#' @details The timestamp is converted to a date as they are only available daily.
#'
#' Category is one of: authPermissions, manyToOneRedirect, notFollowed, notFound,
#'   other, roboted, serverError, soft404.
#'
#'   Platform is one of: mobile, smartphoneOnly or web.
#'
#' @export
#' @family working with search console errors
crawl_errors <- function(siteURL,
                         category="all",
                         platform=c("all","mobile","smartphoneOnly","web"),
                         latestCountsOnly = FALSE) {
  stop("Crawl errors are no longer available in the API")
}

#' Lists a site's sample URLs for crawl errors.
#'
#' @description Category is one of: authPermissions, manyToOneRedirect, notFollowed, notFound,
#'   other, roboted, serverError, soft404.
#'
#'   Platform is one of: mobile, smartphoneOnly or web.
#'
#' @param siteURL The URL of the website to delete. Must include protocol (http://).
#' @param category Crawl error category. Default 'notFound'.
#' @param platform User agent type. Default 'web'.
#'
#' @details
#' See here for details: \url{https://developers.google.com/webmaster-tools/v3/urlcrawlerrorssamples}
#'
#' @return A dataframe of $pageUrl, $last_crawled, $first_detected, $response
#'
#' @export
#' @family working with search console errors
list_crawl_error_samples <- function(siteURL,
                                     category="notFound",
                                     platform="web") {

  siteURL <- check.Url(siteURL, reserved=T)

  ## require pre-existing token, to avoid recursion
  if(is.valid.category.platform(category, platform)) {

    params <- list('category' = category,
                   'platform' = platform)

    lces <-
      googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                     "GET",
                                     path_args = list(sites = "siteURL",
                                                      urlCrawlErrorsSamples = ""),
                                     pars_args = params,
                                     data_parse_function = parse_crawlerror_sample)

    lces(path_arguments = list(sites = siteURL), pars_arguments = params)
  }

}

#' Shows details of errors for individual sample URLs
#'
#' See here for details: https://developers.google.com/webmaster-tools/v3/urlcrawlerrorssamples/get
#'
#' @param siteURL The URL of the website to delete. Must include protocol (http://).
#' @param pageURL A PageUrl taken from list_crawl_error_samples.
#' @param category Crawl error category. Default 'notFound'.
#' @param platform User agent type. Default 'web'.
#'
#' @return Dataframe of $linkedFrom, with the calling URLs $last_crawled, $first_detected and a $exampleURL
#' @family working with search console errors
#' @description
#' pageURL is the relative path (without the site) of the sample URL.
#' It must be one of the URLs returned by list_crawl_error_samples.
#' For example, for the URL https://www.example.com/pagename on the site https://www.example.com/,
#' the url value is pagename (string)
#'
#' Category is one of: authPermissions, manyToOneRedirect, notFollowed, notFound,
#'   other, roboted, serverError, soft404.
#'
#'   Platform is one of: mobile, smartphoneOnly or web.
#'
#' @export
error_sample_url <- function(siteURL,
                             pageURL,
                             category="notFound",
                             platform="web") {

  siteURL <- check.Url(siteURL, reserved = TRUE)
  pageURL <- check.Url(pageURL, checkProtocol = FALSE, reserved = TRUE, repeated = TRUE)


  ## require pre-existing token, to avoid recursion
  if(is.valid.category.platform(category, platform)){

    params <- list('category' = category,
                   'platform' = platform)

    esu <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                          "GET",
                                          path_args = list(sites = "siteURL",
                                                           urlCrawlErrorsSamples = "pageURL"),
                                          pars_args = params,
                                          data_parse_function = parse_errorsample_url)

    esu(path_arguments = list(sites = siteURL,
                              urlCrawlErrorsSamples = pageURL),
        pars_arguments = params)

  }

}

#' Mark As Fixed the individual sample URLs
#'
#' See here for details:
#' https://developers.google.com/webmaster-tools/v3/urlcrawlerrorssamples/markAsFixed
#'
#' @param siteURL The URL of the website to delete. Must include protocol (http://).
#' @param pageURL A PageUrl taken from list_crawl_error_samples.
#' @param category Crawl error category. Default 'notFound'.
#' @param platform User agent type. Default 'web'.
#'
#' @return TRUE if successful, raises an error if not.
#' @family working with search console errors
#'
#' @description
#' pageURL is the relative path (without the site) of the sample URL.
#' It must be one of the URLs returned by list_crawl_error_samples.
#' For example, for the URL https://www.example.com/pagename on the site https://www.example.com/,
#' the url value is pagename (string)
#'
#' Category is one of: authPermissions, manyToOneRedirect, notFollowed, notFound,
#'   other, roboted, serverError, soft404.
#'
#'   Platform is one of: mobile, smartphoneOnly or web.
#'
#' @export
fix_sample_url <- function(siteURL,
                           pageURL,
                           category = "notFound",
                           platform = "web") {

  siteURL <- check.Url(siteURL, reserved = TRUE)
  pageURL <- check.Url(pageURL, checkProtocol = FALSE, reserved = TRUE)

  if(is.valid.category.platform(category, platform)){

    params <- list('category' = category,
                   'platform' = platform)

    fsu <- googleAuthR::gar_api_generator("https://www.googleapis.com/webmasters/v3/",
                                          "DELETE",
                                          path_args = list(sites = "siteURL",
                                                           urlCrawlErrorsSamples = "pageURL"),
                                          pars_args = params)

    fsu(path_arguments = list(sites = siteURL,
                              urlCrawlErrorsSamples = pageURL),
        pars_arguments = params)

    return(TRUE)

  }

  return(FALSE)
}
